### PR TITLE
Dynamic OPUS FEC encoding for WHIP output.

### PR DESCRIPTION
**Pull request by JBRS307**

### CODE REVIEW

File: compositor_pipeline/src/pipeline/encoder/opus.rs [hunk 1]
```diff
@@ -64,6 +63,15 @@ impl AudioEncoder for OpusEncoder {
...
 
+    fn set_packet_loss(&mut self, packet_loss: i32) {
+        match self.encoder.set_packet_loss_perc(packet_loss) {
+            Ok(_) => {}
+            Err(e) => {
+                error!(%e, "Error while setting opus encoder packet loss.");
+            }
+        }
+    }
+
```
Comment (JBRS307) [inline compositor_pipeline/src/pipeline/encoder/opus.rs]: Change that to if let Err()

File: compositor_api/src/output/whip_into.rs [hunk 2]
```diff
@@ -186,30 +186,44 @@ impl TryFrom<WhipOutput> for pipeline::RegisterOutputOptions<output::OutputOptio
...
                                 vec![pipeline::encoder::AudioEncoderOptions::Opus(
                                     opus::OpusEncoderOptions {
                                         channels: resolved_channels.clone().into(),
                                         preset: preset.unwrap_or(OpusEncoderPreset::Voip).into(),
                                         sample_rate: sample_rate.unwrap_or(48000),
-                                        // TODO, options added for compatibility
-                                        forward_error_correction: false,
+                                        forward_error_correction: forward_error_correction
+                                            .unwrap_or(false),
+                                        // Default
```
Comment (wkozyra95) [inline compositor_api/src/output/whip_into.rs]: ?

File: compositor_pipeline/src/pipeline/encoder.rs [hunk 3]
```diff
@@ -257,17 +264,33 @@ where
...
+                packet_loss_sender,
+                config,
+            },
         ))
     }
+
+    fn read_packet_loss(&mut self) -> Option<i32> {
+        let packet_loss_changed = self.packet_loss_receiver.has_changed().unwrap_or(false);
+        if packet_loss_changed {
+            let packet_loss = self.packet_loss_receiver.borrow_and_update().to_owned();
```
Comment (wkozyra95) [inline compositor_pipeline/src/pipeline/encoder.rs]:
```suggestion
            let packet_loss = *self.packet_loss_receiver.borrow();
```

File: compositor_pipeline/src/pipeline/encoder.rs [hunk 4]
```diff
@@ -257,17 +264,33 @@ where
...
             },
-            config,
+            AudioEncoderStreamContext {
+                packet_loss_sender,
+                config,
+            },
         ))
     }
+
+    fn read_packet_loss(&mut self) -> Option<i32> {
```
Comment (wkozyra95) [inline compositor_pipeline/src/pipeline/encoder.rs]:
```suggestion
    fn updated_packet_loss(&mut self) -> Option<i32> {
```
old name implies that it always returns the value

File: compositor_pipeline/src/pipeline/encoder.rs [hunk 5]
```diff
@@ -257,17 +264,33 @@ where
...
+        let packet_loss_changed = self.packet_loss_receiver.has_changed().unwrap_or(false);
+        if packet_loss_changed {
```
Comment (wkozyra95) [inline compositor_pipeline/src/pipeline/encoder.rs]: nit pick, but I use match (more compact)

File: compositor_pipeline/src/pipeline/encoder/fdk_aac.rs [hunk 6]
```diff
@@ -151,6 +151,10 @@ impl AudioEncoder for FdkAacEncoder {
             vec![]
         })
     }
+
+    // For AAC it does completely nothing, however it is needed for trait
```
Comment (wkozyra95) [inline compositor_pipeline/src/pipeline/encoder/fdk_aac.rs]: I don't think this is necessary, if you want to add a comment explaining this, then it should be explanation why the function is empty not why it exists e.g. aac does not have any mechanisms to counteract packet loss

File: compositor_pipeline/src/pipeline/output/whip/setup_track.rs [hunk 7]
```diff
@@ -202,32 +230,117 @@ pub async fn setup_audio_track(
...
+                let remote_inbound_stats = match stats.get(&remote_inbound_id) {
+                    Some(report) => match report {
+                        StatsReportType::RemoteInboundRTP(report) => report,
+                        _ => {
+                            error!("Invalid report type for given key! (This should not happen)");
+                            continue;
+                        }
+                    },
+                    None => {
+                        debug!("OutboundRTP report is empty!");
```
Comment (wkozyra95) [inline compositor_pipeline/src/pipeline/output/whip/setup_track.rs]:
```suggestion
                        debug!("InboundRTP report is empty!");
```

File: compositor_pipeline/src/pipeline/output/whip/setup_track.rs [hunk 8]
```diff
@@ -202,32 +230,117 @@ pub async fn setup_audio_track(
...
+                };
+
+                let remote_inbound_stats = match stats.get(&remote_inbound_id) {
+                    Some(report) => match report {
+                        StatsReportType::RemoteInboundRTP(report) => report,
+                        _ => {
+                            error!("Invalid report type for given key! (This should not happen)");
+                            continue;
+                        }
+                    },
```
Comment (wkozyra95) [inline compositor_pipeline/src/pipeline/output/whip/setup_track.rs]:
```suggestion
                    Some(StatsReportType::RemoteInboundRTP(report) ) =>  report,
                    Some(_) => {
                          error!("Invalid report type for given key! (This should not happen)");
                          continue;
                    },
```

File: compositor_pipeline/src/pipeline/output/whip/setup_track.rs [hunk 9]
```diff
@@ -202,32 +230,117 @@ pub async fn setup_audio_track(
...
                     }
+                };
+
+                let packets_sent: u64 = outbound_stats.packets_sent;
+                // This can be lower than 0 in case of duplicates
+                let packets_lost: u64 = if remote_inbound_stats.packets_lost < 0 {
+                    0
+                } else {
+                    remote_inbound_stats.packets_lost as u64
+                };
```
Comment (wkozyra95) [inline compositor_pipeline/src/pipeline/output/whip/setup_track.rs]:
```suggestion
                let packets_lost: u64 = i64::max(remote_inbound_stats.packets_lost, 0) as u64
```

**Review by wkozyra95 — APPROVED** [review]
